scripts/scriptLoader.lua

```diff
@@ -0,0 +1,107 @@
+--ScriptLoader Singleton
+---@class ScriptLoader
+local ScriptLoader = {
+    ScriptData = {
+        GeneratedScriptIds = {}, ---@type table<string, string> Mapping of filepath to script Id
+    },
+}
+local template = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
+
+---@param filePath string File path to load
+---@return string|nil ScriptID
+function ScriptLoader.generateScriptId(filePath)
+    if not string.match(filePath, "%S") then
+        return "" -- should throw, this is bad if we fail this?
+    end
+
+    local seed = 0
+    for i = 1, #filePath do
+        local charCode = string.byte(filePath:normalizePath(), i)
+        seed = seed + charCode
+    end
+    math.randomseed(seed)
+
+    local scriptID = template:gsub("x", function() return string.format("%x", math.random(0, 15)) end)
+    ScriptLoader.ScriptData.GeneratedScriptIds[filePath] = scriptID
+    tes3mp.LogMessage(enumerations.log.VERBOSE, '[customEventHooks]: Generated ScriptID for script: "' .. filePath..'" is ' .. scriptID)
+
+    return scriptID
+end
+
+function ScriptLoader.isScriptLoaded(filePath)
+    if package.loaded[filePath] then
+        return true
+    end
+    return false
+end
+
+---@param filePath string
+---@return boolean isValid If script can be loaded
+---@return string|nil errorMessage error message for why script wasn't loaded
+function ScriptLoader.canLoadScript(filePath)
+    if package.loaded[filePath] then
+        return false, "Package already laoded"
+    end
+
+    return true, nil
+end
+
+---@param filePath string
+---@return boolean Result is successful
+function ScriptLoader.loadScript(filePath)
+    local result = prequire(filePath)
+    if result then
+        -- TODO: implement
+        eventHandler.OnScriptLoad(filePath)
+        return true
+    else
+        return false
+    end
+end
+
+function ScriptLoader.getScriptId(filePath)
+    if not filePath then
+        return ""
+    end
+
+    local filePath = filePath:normalizePath()
+
+    if ScriptLoader.ScriptData.GeneratedScriptIds[filePath] then
+        tes3mp.LogMessage(enumerations.log.VERBOSE, '[ScriptLoader][getScriptID]: ScriptID for script: "' .. filePath..'" is ' .. ScriptLoader.ScriptData.GeneratedScriptIds[filePath])
+        return ScriptLoader.ScriptData.GeneratedScriptIds[filePath]
+     else
+        tes3mp.LogMessage(enumerations.log.VERBOSE, '[ScriptLoader]: ScriptID not found for script: "' .. filePath..'"')
+        return ""
+     end
+end
+
+-- TODO: figure out wtf this was supposed to be doing
```

Comment:
Im not 100% sure how scripts were being loaded/unloaded; this was a copy paste of the previous logic with most of the not-unloading stripped out

Reply:
wouldnt it make more send for 72 to return nil if its not found instead of an empty string?

Reply:
It's confusing because this isn't actually the UN-loading logic, this is for RE-loading; old function pointers persist when you do a require on top of an old one, so this section nullifies the refs already existing and then places the new ones on top.
The actual UN-loading logic is simply this:
```
                -- Fully unload the script by setting it to nil
                package.loaded[scriptName] = nil```
                
                Now that I'm looking at it, though, unloading is probably not complete as it should also do this to nullify the old refs as reloading does. If you /unload, then /load a script with the current version, it'll have the old problem of functions simply not being updated.

```
               for key, value in pairs(package.loaded[scriptName]) do
                        package.loaded[scriptName][key] = nil
                end
```